File: services/backend/pkg/models/vuln_kb.go
package models

import (
	"time"

	"github.com/uptrace/bun"
)

type VulnKBEntry struct {
	bun.BaseModel `bun:"table:vuln_kb"`

	VulnID            string        `bun:"vuln_id,type:text,pk" json:"vuln_id"`
	Description       string        `bun:"description,type:text,default:''" json:"description"`
	Severity          string        `bun:"severity,type:text,default:''" json:"severity"`
	CVSSVector        string        `bun:"cvss_vector,type:text,default:''" json:"cvss_vector"`
	CVSSScore         float64       `bun:"cvss_score,type:float,default:0" json:"cvss_score"`
	PublishedDate     *time.Time    `bun:"published_date,type:timestamptz" json:"published_date"`
	ModifiedDate      *time.Time    `bun:"modified_date,type:timestamptz" json:"modified_date"`
	References        []KBRef       `bun:"references,type:jsonb" json:"references"`
	ExploitAvailable  bool          `bun:"exploit_available,type:bool,default:false" json:"exploit_available"`
	FetchedAt         time.Time     `bun:"fetched_at,type:timestamptz,default:now()" json:"fetched_at"`
	HistoryEventCount int           `bun:"-" json:"history_event_count,omitempty"`
	LastChangeAt      *time.Time    `bun:"-" json:"last_change_at,omitempty"`
	LastChange        *KBLastChange `bun:"-" json:"last_change,omitempty"`
}

type KBRef struct {
	URL    string `json:"url"`
	Source string `json:"source"`
}

type KBLastChange struct {
	EventName        string    `json:"event_name"`
	Source           string    `json:"source"`
	SourceIdentifier string    `json:"source_identifier"`
	ObservedAt       time.Time `json:"observed_at"`
}

File: services/frontend/lib/api/types/vulnkb.ts
export interface VulnKBEntry {
  vuln_id: string;
  description: string;
  severity: string;
  cvss_vector: string;
  cvss_score: number;
  published_date: string | null;
  modified_date: string | null;
  references: { url: string; source: string }[];
  exploit_available: boolean;
  fetched_at: string;
  history_event_count?: number;
  last_change_at?: string | null;
  last_change?: {
    event_name: string;
    source: string;
    source_identifier: string;
    observed_at: string;
  } | null;
}

export interface PublicVulnerabilityIntelligenceHistory {
  data: import('./vulnerability-intelligence').VulnerabilityIntelligenceChangeEvent[];
  total: number;
  has_more: boolean;
  next_before_at?: string | null;
  next_before_id?: string | null;
}

export interface VulnerabilityExposurePosture {
  state: string;
  cve_state: string;
  severity: string;
  cvss_score: number;
  reason: string;
  observed_at?: string | null;
  change_event_id?: string | null;
  updated_at?: string | null;
}

export interface VulnerabilityExposureRow {
  finding_id: string;
  scan_id: string;
  image_name: string;
  image_tag: string;
  completed_at?: string | null;
  package_name: string;
  installed_version: string;
  fixed_version: string;
  scan_severity: string;
  scan_cvss_score: number;
  posture?: VulnerabilityExposurePosture | null;
}

export interface VulnerabilityExposureSummary {
  findings: number;
  scans: number;
  changed: number;
  needs_rescan: number;
  fix_available: number;
}

export interface VulnerabilityExposureResponse {
  data: VulnerabilityExposureRow[];
  total: number;
  page: number;
  limit: number;
  has_more: boolean;
  summary: VulnerabilityExposureSummary;
}
